synapse/handlers/appservice.py

```diff
@@ -307,6 +314,23 @@ async def _notify_interested_services_ephemeral(
                             service, "presence", new_token
                         )
 
+                    elif (
+                        stream_key == "to_device_key"
+                        and self.msc2409_to_device_messages_enabled
+                    ):
+                        # Retrieve an iterable of to-device message events, as well as the
+                        # maximum stream token of the messages we were able to retrieve.
+                        events = await self._handle_to_device(service, new_token, users)
+                        if events:
+                            self.scheduler.submit_ephemeral_events_for_as(
+                                service, events
+                            )
+
+                        # Persist the latest handled stream token for this appservice
+                        await self.store.set_type_stream_id_for_appservice(
+                            service, "to_device", new_token
+                        )
```

Comment:
I haven't chased through the code: once the events are sent to  `submit_ephemeral_events_for_as`, will they reliably get delivered?

Reply:
it's probably not worth blocking right now. Suggest you add some FIXME-style comments here to note the outstanding work.

Reply:
Good idea. I've added comments for the following outstanding items from this PR in c691ef0992da5ad3b4998a698a39b97acfb81aea:

* adding constants for the various stream keys
* fixing a user being either a `str` or a `UserID` when being passed to `notify_interested_services_ephemeral`

There's some existing issues for the outstanding work as well:

* EDU -> AS reliability (https://github.com/matrix-org/synapse/issues/11150)
* deleting to-device messages intended for exclusive namespace users after they've been successfully delivered to the AS (https://github.com/matrix-org/synapse/issues/11423)
* adding logic for not spamming AS's with historical to-device events when they're first connected (https://github.com/matrix-org/synapse/issues/10836)